RavenDB-22986 Optimizing free space allocation

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22986 

### Additional description

Performance optimization, further vectorizing the finding of available free space.

ayende committed:
RavenDB-22986 - Refactoring free space handling to be:
* Simpler code
* Dedicated code paths (for ease of inlining, etc) for the various sizes

ayende committed:
RavenDB-22986
- Vectorized more of StreamBitArray
- Doing _evil_ things to get high performance

redknightlois commented on src/Voron/Impl/FreeSpace/StreamBitArray.cs:
```diff
@@ -84,88 +86,177 @@ private void Write(byte* ptr)
+        {
+            var aligned = Vector256.LoadUnsafe(ref _inner[0], (nuint)i);
+            var unaligned = Vector256.LoadUnsafe(ref shiftedInner, (nuint)i);
+
```
If we are sure this is aligned we could use `Vector256.LoadAligned` instead.

ayende replied:
No, this will require us to be aligned on 32 bytes boundary.
And we have no way to require that, I'm afraid.

redknightlois commented on src/Voron/Impl/FreeSpace/StreamBitArray.cs:
```diff
@@ -84,88 +86,177 @@ private void Write(byte* ptr)
+                NextElement:
+                while (idx < i + Vector256<uint>.Count && 
+                       _inner[idx] != uint.MaxValue)
+                {
```
Used like this, there is no way to elide the `_inner` bound checks.

ayende replied:
`_inner` isn't an array, there aren't any bound checks here, see:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACAVwDtdsAzGB3DKJsBgYBhBgG8aDKQ0nT6DDgEsEMACYNFLIQEkMMfLgDaADgC6AbllSrDeZqEBZABT2G2AJTib02wHYGuvpG2BY2AL40YUA==

ayende committed:
RavenDB-22986
- Remember the maximum number of pages that we _couldn't_ allocate from a segment, and skip checking it if this is known to not have the required consecutive range
- On FreePage, we simply delete the information for that segment (requires re-check and then we'll remember if needed)
- On RollBack, we all information and require a re-check (could make that into a proper state on the env record, didn't do so far to make it easier to merge to prev versions)

- In this way, this keeps minimal state, and it is basically self-healing if needed

grisha-kotler commented on src/Voron/Impl/FreeSpace/FreeSpaceHandling.cs:
```diff
@@ -184,15 +189,26 @@ public void Clear()
             do
             {
+                long currentSectionId = it.CurrentKey;
+                if(_maxConsecutiveRange.TryGetValue(currentSectionId, out var knownMax) && knownMax >= num)
```
The condition here is wrong - should be `knownMax <= num`.
